process/release_process.md

```diff
@@ -111,6 +111,31 @@ It leads us to need a separate 'xt-core' repository that will be the entry point
     - Merging the 'development' branch into the 'main' branch
     - Adding the commit that locks all the dependencies
     - Assigning the release tag to the HEAD commit of the 'main' branch
+- **Approach #3**:
+   - Development is done in the 'development' branch. In this branch, all dependencies can stick to the HEAD. In this stage we separate repositories that are including external dependencies and requires actual commit to stabilize external revision. They could be:
+     - Moulin meta repository which includes moulin.yaml
+     - Zephyr based repository with west.yaml which is the dependency for Moulin repository
+   - Release is done by creating a separate branch for each release on top-level repositories with the name of current release
+   - Version format as **<component>-vX.X.X** is acceptable because
+   - Steps to create release are the following:
+     - On repo with external dependensies (moulin or west.yaml) create a separate branch with the release name
+     - Make a stabilizaition commit with the following changes: all external components switched from branch to specific revision.
+     - Repeat this steps for all repo with external dependencies.
+   - Procs:
+     - Release process results the following **git log** for repo with external dependencies:
+       ```
+               v0.1.0              v0.1.1
+                / *                / *
+            *- * - * - *  - * - * - * - * (master)
+       ```
```

Comment:
Consider the following scenario:

1. You make v0.1.0 release
2. Development continues in master branch, new experimental/unstable/etc features are added
3.  Customer have found issue in v0.1.0 release and you need to make a hotfix
4. You create hotfix commit and you need to make a new "patch" release v0.1.0. Which commit you will use as a  base for this release? You can't use "master" branch because it now includes new unstable features.

Reply:
I would make v0.1.1

Reply:
Yes, I understand. But at which commit it will point? Consider this:
```
 Master branch                             | v0.1.0 release branch
 - A: last feature before release          |
 \ -----------------------------------------> B: Release commit
 - C: New feature 1                        |
 - B: New feature 2			   |
 - D: Bugfix for rel 0.1.0                 |
 ```

How do you want to move forward? I see two options. Option A is to branch at commit `D`:

```
 Master branch                             | v0.1.0 release branch | v0.1.1 release branch
 - A: last feature before release          |                       |
 \ -----------------------------------------> B: Release commit    |
 - C: New feature 1                        |                       |
 - B: New feature 2			   |                       |
 - D: Bugfix for rel 0.1.0                 |                       |
 \------------------------------------------------------------------> E: Release commit
```

Option B is to backport fix `D`, branching from `B` and make new release:

```
 Master branch                              v0.1.0 release branch   v0.1.1 release branch
 - A: last feature before release
 \ -----------------------------------------> B: Release commit
 - C: New feature 1                           |
 - B: New feature 2                           |
 - D: Bugfix for rel 0.1.0                    |
                                              \----------------------> E: Cherry-pick of D
 					                               F: Release commit
```


Which one are you proposing?

Reply:
Minor comment from my side.

According to the semversion documentation:

https://semver.org/#how-do-i-know-when-to-release-100

```
How do I know when to release 1.0.0?
If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backward compatibility, you should probably already be 1.0.0.
```

**=> I suggest mentioning v1.x.y in the documentation, as many of the developed components are old enough to be outside of the 'initial development phase'.**

Reply:
I do force-push with changing commit order. 
Again please note that "That's why I propose separate approach because I believe there is no generic approach"
And it works in our case with a small team working fast on some branch and symultaniously imolementing cross-dependent features.

Reply:
Force-push into public branches, that are used by multiple developers, is so big no-no that I see no sense in discussion this approach further.  

Here is my `NACK`.

Please provide alternative solution.

Reply:
> This approach where tested in aos development process. IIRC because of the lack of time we've merged fixups on commits and then made a cleanup. 

Okay, I see sense in creating development branch, making a bunch of fast and dirty commits and then rearranging them into new clean and stable branch. But this implies that you have some "stable" main branch where day-to-day development happens. And there is someone who is responsible on moving patches from dev branch to the main one. 

But making releases directly from a development branch that can be force-pushed later... I just don't get it.

Reply:
All rearangements will not touch already released points. releases are releases. Let's say this way. My vision is the following:
         v0.1.0
         /*
 * - * - * - * - * - * - *
 [no-no][can be rearranged]
 so techically we have master and development symbiosis in one branch

Reply:
You can't rearrange commits without risking introducing issues. 

Basically you proposing to rearrange this:

```
 Master branch                              v0.1.0 release branchh
 - A: last feature before release
 \ -----------------------------------------> B: Release commit
 - C: New feature 1
 - D: New feature 2
 - E: Bugfix for rel 0.1.0
```

Into this:


```
 Master branch                              v0.1.0 release branch   v0.1.1 release branch
 - A: last feature before release
 \ -----------------------------------------> B: Release commit
 - E: Bugfix for rel 0.1.0                    |
 \------------------------------------------------------------------> F:Release commit
 - C: New feature 1                           |
 - D: New feature 2                           |
```

But what if `E` already depends on `C` ? Or `C` depends directly on `A`? Person who does rebase will get a conflict during rebase. They will be forced to rework both `E` and `C` (and maybe `D` as well). But these patches were reviewed already. Then we either need a new round of review for all the affected patches or we need to trust that person who makes the rebase will not introduce new bugs (whis is not an option).  

In our current scheme, we backport `E` onto release branch, fixing any conflicts and then review it as a new patch for release branch. But this is one isolated commit, not the whole history. You are proposing to either review all commits as new or to not review them at all, risking introducing new bugs into already working and tested code.

Also, what to do when you made 2.0.0 release and suddenly you need to update v.0.1.0 due to some critical issue? You can't rearrange this part anymore...

Also, forced pushes will mess with CI, with other developers (not all external users are sitting on releases),  etc.

I don't believe that this development process is sustainable. @GrygiriiS, what is your opinion on this?

Reply:
> Okay. In this case we backport E to the release branch. Then make v0.1.1. After that, after finishing all features we want to release v0.1.2. 

By Semver definition, v0.1.2 is "PATCH version when you make backward compatible bug fixes". You can't add new features here. Please read Semver spec, it is short.

> And in this case we receive a great pain with pushing all featres to v0.1 branch (remember - e is already backported and changes). 

You  don't need to add any new features to v0.1 branch. Only bugfixes. 

> So in release v0.1.1 we have already backported E with parts of C and D. So we have to update both C and D and rereview. So we do the same thing but twice

Nope, this may be needed only for v.0.2.0 release, for which you create a new branch of top of master. And of course in this case you don't need to rebase anything. 

> I will repeat again- this approach work when we need to maximize speed and minimize merging overhead when we work on some feature in a small team which does many cross-referenced features.

This will impair quality. Everyone is okay with increased speed. But not by sacrificing quality.  Instead of rushing things it is better to postpone release but make it good. @klogg already told this and I believe he can confirm this again.